File: backend/apps/communication/models/event.py
from django.conf import settings
from django.db import models

from apps.core.models.mixins import TenantMixin, TimestampMixin


class Event(TenantMixin, TimestampMixin):
    class EventType(models.TextChoices):
        PRIVATE = "Private", "Private"
        PUBLIC = "Public", "Public"

    subject = models.CharField(max_length=255)
    description = models.TextField(blank=True)
    location = models.CharField(max_length=500, blank=True)
    starts_on = models.DateTimeField()
    ends_on = models.DateTimeField(null=True, blank=True)
    all_day = models.BooleanField(default=False)
    event_type = models.CharField(
        max_length=20, choices=EventType.choices, default=EventType.PRIVATE
    )
    color = models.CharField(max_length=20, blank=True)

    # Link to entity
    entity_type = models.CharField(max_length=50, blank=True)  # "lead", "deal"
    entity_id = models.IntegerField(null=True, blank=True)

    owner = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="events",
    )

    class Meta:
        db_table = "crm_events"
        ordering = ["starts_on"]

    def __str__(self):
        return self.subject


class EventParticipant(TenantMixin):
    class AttendingStatus(models.TextChoices):
        YES = "Yes", "Yes"
        NO = "No", "No"
        MAYBE = "Maybe", "Maybe"

    event = models.ForeignKey(
        Event, on_delete=models.CASCADE, related_name="participants"
    )
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        null=True,
        blank=True,
    )
    email = models.EmailField()
    attending = models.CharField(
        max_length=10,
        choices=AttendingStatus.choices,
        default=AttendingStatus.YES,
    )

    class Meta:
        db_table = "crm_event_participants"

    def __str__(self):
        return f"{self.email} – {self.attending}"

File: frontend/src/types/event.ts
export interface CalendarEvent {
  id: number
  subject: string
  description: string
  location: string
  starts_on: string
  ends_on: string | null
  all_day: boolean
  event_type: "Private" | "Public"
  color: string
  entity_type: string
  entity_id: number | null
  owner_name: string
  owner_email: string
  participants: EventParticipant[]
  created_at: string
}

export interface EventParticipant {
  id: number
  user_id: number | null
  email: string
  attending: "Yes" | "No" | "Maybe"
}

export interface EventCreate {
  subject: string
  description?: string
  location?: string
  starts_on: string
  ends_on?: string
  all_day?: boolean
  event_type?: string
  color?: string
  entity_type?: string
  entity_id?: number
  participants?: { email: string }[]
}

export interface EventUpdate extends Partial<EventCreate> {}
